include/circt/Dialect/Calyx/CalyxStructure.td

```diff
@@ -76,9 +75,29 @@ def ComponentOp : CalyxOp<"component", [
       /// Hooks for the input/output type enumeration in FunctionLike.
       unsigned getNumFuncArguments() { return getType().getNumInputs(); }
       unsigned getNumFuncResults() { return getType().getNumResults(); }
+
+      /// This is a helper function that should only be used to get the WiresOp or
+      /// ControlOp of a ComponentOp, which are guaranteed to exist and generally at
+      /// the end of a component's body. In the worst case, this will run in linear
+      /// time with respect to the number of instances within the cell.
+      template <typename Op>
+      Operation* getOpWithType() {
+        auto body = &getOperation()->getRegion(0).front();
+        auto opIt = llvm::find_if(llvm::reverse(*body),
```

Comment:
There's already a templated `getOps<OpType>()` that will give you an iterator over ops of a specific type: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Block.h#L185. Might clean this up a bit. IMO iterating in reverse seems like an optimization you could apply later if you really find iterating in forward order to be a major performance bottleneck.